## Big Data Assignment 5 from Team dunnhumby

### import all libraries

In [164]:
import geocoder
import pandas as pd
from geopy.distance import great_circle

### read dictionary of capitals of US 50 states

In [165]:
capitals = {'Alabama': 'Montgomery', 'Alaska': 'Juneau', 'Arizona': 'Phoenix',
   'Arkansas': 'Little Rock', 'California': 'Sacramento', 'Colorado': 'Denver',
   'Connecticut': 'Hartford', 'Delaware': 'Dover', 'Florida': 'Tallahassee',
   'Georgia': 'Atlanta', 'Hawaii': 'Honolulu', 'Idaho': 'Boise', 'Illinois':
   'Springfield', 'Indiana': 'Indianapolis', 'Iowa': 'Des Moines', 'Kansas':
   'Topeka', 'Kentucky': 'Frankfort', 'Louisiana': 'Baton Rouge', 'Maine':
   'Augusta', 'Maryland': 'Annapolis', 'Massachusetts': 'Boston', 'Michigan':
   'Lansing', 'Minnesota': 'Saint Paul', 'Mississippi': 'Jackson', 'Missouri':
   'Jefferson City', 'Montana': 'Helena', 'Nebraska': 'Lincoln', 'Nevada':
   'Carson City', 'New Hampshire': 'Concord', 'New Jersey': 'Trenton', 'New Mexico': 'Santa Fe', 'New York': 'Albany', 'North Carolina': 'Raleigh',
   'North Dakota': 'Bismarck', 'Ohio': 'Columbus', 'Oklahoma': 'Oklahoma City',
   'Oregon': 'Salem', 'Pennsylvania': 'Harrisburg', 'Rhode Island': 'Providence',
   'South Carolina': 'Columbia', 'South Dakota': 'Pierre', 'Tennessee':
   'Nashville', 'Texas': 'Austin', 'Utah': 'Salt Lake City', 'Vermont':
   'Montpelier', 'Virginia': 'Richmond', 'Washington': 'Olympia', 'West Virginia': 'Charleston', 'Wisconsin': 'Madison', 'Wyoming': 'Cheyenne'}

### get the latitudes and longtitudes

In [168]:
cities=[]
def getlatlng(state, city):
    name = city+", "+state
    g = geocoder.google(name).latlng
    if g is not None:
        cities.append([city, (g[0],g[1])])
    else:
        getlatlng(state, city)       

In [169]:
for state, city in capitals.items():
    getlatlng(state, city)

### get distiant in miles between any two cities

In [170]:
r1=[]
for i in cities:
    for j in cities:
        dist = great_circle(i[1], j[1]).miles
        if dist>0 and dist<500:
            r1.append([i[0],j[0],dist])

In [171]:
r1

[['Montgomery', 'Little Rock', 382.0732194127232],
 ['Montgomery', 'Tallahassee', 178.74694476583971],
 ['Montgomery', 'Atlanta', 146.2518406585516],
 ['Montgomery', 'Frankfort', 411.14891458522874],
 ['Montgomery', 'Baton Rouge', 314.45152193195923],
 ['Montgomery', 'Jackson', 226.9003943837123],
 ['Montgomery', 'Raleigh', 497.79056264400265],
 ['Montgomery', 'Columbia', 324.755707550272],
 ['Montgomery', 'Nashville', 263.7795549804174],
 ['Montgomery', 'Charleston', 489.92027243527076],
 ['Phoenix', 'Santa Fe', 381.86079992040425],
 ['Little Rock', 'Montgomery', 382.07321941272323],
 ['Little Rock', 'Atlanta', 456.52299623814434],
 ['Little Rock', 'Springfield', 377.01742490419144],
 ['Little Rock', 'Indianapolis', 483.7215732358524],
 ['Little Rock', 'Des Moines', 479.079526861448],
 ['Little Rock', 'Topeka', 352.0692163615059],
 ['Little Rock', 'Frankfort', 476.0887664241285],
 ['Little Rock', 'Baton Rouge', 303.82157521211974],
 ['Little Rock', 'Jackson', 208.13281695155183],
 ['L

### get the result

In [178]:
## start from two cities; and add one more city for each loop and check if the current distant <1000
## stop the do-while loop when there is no result in the list (<1000)
result = r1
flag = 1
while flag>0:
    new_result=[]
    for i in result:
        for j in r1:
            if i[-1]+j[-1]<1000 and i[-2]==j[0] and j[-2] not in i:
                p=i[0:-1]
                new = ((j[1],i[-1]+j[-1]))
                p.extend(new)
                new_result.append(p)
    last_result = result
    result = new_result
    flag  = len(result)

In [179]:
(last_result)

[['Hartford',
  'Albany',
  'Montpelier',
  'Concord',
  'Boston',
  'Providence',
  'Trenton',
  'Dover',
  'Annapolis',
  'Richmond',
  'Raleigh',
  996.4352314804204],
 ['Hartford',
  'Albany',
  'Montpelier',
  'Concord',
  'Boston',
  'Providence',
  'Trenton',
  'Harrisburg',
  'Dover',
  'Annapolis',
  'Richmond',
  992.4157459382996],
 ['Hartford',
  'Providence',
  'Boston',
  'Augusta',
  'Concord',
  'Montpelier',
  'Albany',
  'Trenton',
  'Dover',
  'Annapolis',
  'Harrisburg',
  993.0878276795238],
 ['Hartford',
  'Providence',
  'Boston',
  'Concord',
  'Augusta',
  'Montpelier',
  'Albany',
  'Trenton',
  'Dover',
  'Annapolis',
  'Harrisburg',
  956.2571280291061],
 ['Hartford',
  'Providence',
  'Boston',
  'Concord',
  'Augusta',
  'Montpelier',
  'Albany',
  'Trenton',
  'Dover',
  'Annapolis',
  'Richmond',
  976.1309912632727],
 ['Hartford',
  'Providence',
  'Boston',
  'Concord',
  'Augusta',
  'Montpelier',
  'Albany',
  'Trenton',
  'Harrisburg',
  'Dover',
  

In [180]:
len(last_result)

478

In [181]:
len(last_result[0])-1

11

### Conclusion

Within a distant limitation of 1000 miles, the helicoptor is able to fly through 11 capitals of US states at most. There are 478 possible routes as shown above.